include/toypp/sharedptr.hpp

```diff
@@ -0,0 +1,224 @@
+#ifndef TOYPP_SHAREDPTR_HPP_
+#define TOYPP_SHAREDPTR_HPP_
+
+#include <atomic>
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+namespace tpp {
+
+namespace details {
+template <typename T>
+struct is_unbounded_array : std::false_type {};
+
+template <typename T>
+struct is_unbounded_array<T[]> : std::true_type {};
+}  // namespace details
+
+template <typename T>
+class WeakPtr;
+
+// TODO support static pointer cast conversions
+template <typename T>
+class SharedPtr {
+  using value_type = std::remove_cv_t<std::remove_reference_t<T>>;
+
+  static constexpr auto k_null_deleter = [](value_type*) { /* do nothing. */ };
+  static constexpr auto k_default_deleter = [](value_type* ptr) { delete ptr; };
+
+  struct Shared {
+    std::atomic<std::size_t> ptr_refcount = 1;
+    std::atomic<std::size_t> shared_refcount = 1;
+    std::function<void(value_type*)> deleter = k_default_deleter;
+  };
+
+  inline static Shared s_null_shared{.ptr_refcount = 0, .shared_refcount = 1, .deleter = k_null_deleter};
+
+  value_type* _ptr;
+  Shared* _shared;
+
+  friend class WeakPtr<T>;
+  friend class WeakPtr<value_type>;
+
+  void destroy() {
+    if (!_shared) {
+      return;
+    }
+    if (_shared->ptr_refcount.fetch_sub(1) == 1) {
+      _shared->deleter(_ptr);
+    }
+    if (_shared->shared_refcount.fetch_sub(1) == 1) {
+      delete _shared;
+    }
+  }
+
+  explicit SharedPtr(value_type* ptr, Shared* shared) noexcept : _ptr(ptr), _shared(shared) {}
+
+ public:
+  SharedPtr() noexcept : _ptr(nullptr), _shared(&s_null_shared) {
+    _shared->ptr_refcount.fetch_add(1);
+    _shared->shared_refcount.fetch_add(1);
+  }
+  SharedPtr(value_type* ptr) : _ptr(ptr), _shared(new Shared{}) {}
+
+  template <typename F>
+  SharedPtr(value_type* ptr, F&& deleter) : _ptr(ptr), _shared(new Shared{.deleter = std::forward<F>(deleter)}) {}
+
+  SharedPtr(const SharedPtr& other) : _ptr(other._ptr), _shared(other._shared) {
+    other._shared->ptr_refcount.fetch_add(1);
+    other._shared->shared_refcount.fetch_add(1);
+  }
+  SharedPtr(SharedPtr&& other) noexcept
+    : _ptr(std::exchange(other._ptr, nullptr))
+    , _shared(std::exchange(other._shared, nullptr))
+  {}
+
+  SharedPtr& operator=(const SharedPtr& other) {
+    if (this == &other) {
+      return *this;
+    }
+    destroy();
+    _ptr = other._ptr;
+    _shared = other._shared;
+    other._shared->ptr_refcount.fetch_add(1);
+    other._shared->shared_refcount.fetch_add(1);
+    return *this;
+  }
+  SharedPtr& operator=(SharedPtr&& other) noexcept {
+    std::swap(_ptr, other._ptr);
+    std::swap(_shared, other._shared);
+    return *this;
```

Comment:
The move assignment operator should not swap the values.  It should leave `other` in an empty state.  While swapping technically satisfies the requirement that `other` is left in a valid but unspecified state, this will be surprising to users.

Reply:
@dkolsen-pgi do you recommend calling `destroy` which potentially will call `delete` here? I wanted to make sure it's both fast (avoids unnecessary operations) and has no room for exceptions.